tests/test_z3_visitor.py

```diff
@@ -41,16 +30,127 @@ def f(x: bool, y: bool, z: bool, a):
         '''
         pass
     """
+]
+
+# test cases for container (list/set/tuple) expressions
+container_list = [
+    """
+    def in_list(x: int):
+        '''
+        Preconditions:
+            - x in [1, 2, 3]
+        '''
+        pass
+    """,
+    """
+    def in_set(x: int):
+        '''
+        Preconditions:
+            - x in {1, 2, 3}
+        '''
+        pass
+    """,
+    """
+    def in_tuple(x: int):
+        '''
+        Preconditions:
+            - x in (1, 2, 3)
+        '''
+        pass
+    """,
+    """
+    def not_in_list(x: int):
+        '''
+        Preconditions:
+            - x not in [1, 2, 3]
+        '''
+        pass
+    """,
+    """
+    def not_in_set(x: int):
+        '''
+        Preconditions:
+            - x not in {1, 2, 3}
+        '''
+        pass
+    """,
+    """
+    def not_in_tuple(x: int):
+        '''
+        Preconditions:
+            - x not in (1, 2, 3)
+        '''
+        pass
+    """,
+    """
+    def in_empty_container(x: int):
+        '''
+        Preconditions:
+            - x in []
+            - x in set()
+            - x in ()
+        '''
+        pass
+    """,
+    """
+    def not_in_empty_container(x: int):
+        '''
+        Preconditions:
+            - x not in []
+            - x not in set()
+            - x not in ()
+        '''
+        pass
+    """,
+]
+
+
+# expected arithmetic expressions
+x = z3.Int("x")
+y = z3.Int("y")
+z = z3.Real("z")
+arithmetic_expected = [[x**2 + y**2 == z**2, z3.And([x > 0, y > 0, z == 0])]]
+
+# expected boolean expressions
+x = z3.Bool("x")
+y = z3.Bool("y")
+z = z3.Bool("z")
+boolean_expected = [[z3.And([x, y, z]), z3.Not(z3.Or([x, y, z]))]]
+
+# expected container expressions
+x = z3.Int("x")
+container_expected = [
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.Or(x == 1, x == 2, x == 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.And(x != 1, x != 2, x != 3)],
+    [z3.BoolVal(False), z3.BoolVal(False)],
```

Comment:
First, each of these should be in their own function (as you've done for all of the above test cases in `container_expected`). Same with the "not in empty" test below.

Second (more as a comment), I'm surprised that there are three precondition expressions but only two Z3 expressions here. Can you explain why that is?

Reply:
Hi Professor, sorry that I wan't quite clear what do you mean by the first comment. I didn't see how the test cases for empty containers differ from the ones above.

About the second problem, sorry that I wasn't being careful enough. The precondition for empty set turns out to be ignored since PythonTA couldn't infer the return type of `set()` function and skipped creating z3 expression for this precondition. However, I cannot fix this problem with modifications in `ExprWrapper` as the problem lies in the `parse_assertions` function in `contracts/__init__.py`.  Currently we may not be able to create the test cases with empty sets.

Reply:
@Raine-Yang-UofT sure, what I meant was that I'd like to see each precondition as a separate test case, i.e., splitting up the `x in []` and `x in ()` preconditions into separate test cases (and function). This aligns with the other test cases you've added, which each have one precondition per function.

Good point about the usage of `set()`. It's fine to not include test cases for empty sets in this PR.

Reply:
@david-yz-liu Thanks, I have updated the tests. In addition, perhaps I can submit a issue concerning the problem with empty sets.